File: backend/app/models/action_pack.py
"""Evidence-first Action Pack contracts used by Gemini and the API."""

from __future__ import annotations

from datetime import date, datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.generated_asset import GeneratedAsset

OutputOption = Literal[
    "complete_action_pack",
    "quick_revision_notes",
    "visual_mind_map",
    "voice_lesson",
    "flashcards",
    "priority_coding_problems",
    "interview_revision_sheet",
]


class EvidenceReference(BaseModel):
    model_config = ConfigDict(extra="forbid")

    resource_id: UUID
    title: str
    location: str | None = None
    confidence: float = Field(ge=0, le=1)
    basis: Literal["source_derived", "ai_inferred"]
    support: str


class ResourceChoice(BaseModel):
    model_config = ConfigDict(extra="forbid")

    resource_id: UUID
    title: str
    reason: str
    evidence: list[EvidenceReference] = Field(min_length=1)


class BacklogReduction(BaseModel):
    model_config = ConfigDict(extra="forbid")

    resource_count: int = Field(ge=2)
    estimated_original_minutes: int | None = Field(default=None, ge=0)
    repeated_content_percentage: float = Field(ge=0, le=100)
    metric_methodology: str
    essential_resources: list[ResourceChoice]
    optional_resources: list[ResourceChoice]
    skippable_resources: list[ResourceChoice]


class StartHere(BaseModel):
    model_config = ConfigDict(extra="forbid")

    topic_or_resource: str
    why: str
    estimated_minutes: int | None = Field(default=None, ge=0)
    evidence: list[EvidenceReference] = Field(min_length=1)


class CommonTopic(BaseModel):
    model_config = ConfigDict(extra="forbid")

    topic: str
    explanation: str
    source_count: int = Field(ge=2)
    evidence: list[EvidenceReference] = Field(min_length=2)


class UniqueInsight(BaseModel):
    model_config = ConfigDict(extra="forbid")

    insight: str
    why_it_matters: str
    evidence: list[EvidenceReference] = Field(min_length=1)


class ConflictSide(BaseModel):
    model_config = ConfigDict(extra="forbid")

    position: str
    evidence: list[EvidenceReference] = Field(min_length=1)


class Contradiction(BaseModel):
    model_config = ConfigDict(extra="forbid")

    topic: str
    sides: list[ConflictSide] = Field(min_length=2)
    neutral_explanation: str
    recommendation: str | None
    recommendation_confidence: float | None = Field(default=None, ge=0, le=1)


class ResourceVerdict(BaseModel):
    model_config = ConfigDict(extra="forbid")

    resource_id: UUID
    title: str
    verdict: Literal[
        "essential",
        "use_selected_sections",
        "reference_only",
        "safe_to_skip",
        "unavailable_or_low_confidence",
    ]
    reason: str
    selected_sections: list[str]
    evidence: list[EvidenceReference] = Field(min_length=1)


class MergedNote(BaseModel):
    model_config = ConfigDict(extra="forbid")

    topic: str
    concise_notes: list[str] = Field(min_length=1)
    syntax_or_pseudocode: str | None
    recognition_clues: list[str]
    common_mistakes: list[str]
    memory_cues: list[str]
    evidence: list[EvidenceReference] = Field(min_length=1)


class PriorityProblem(BaseModel):
    model_config = ConfigDict(extra="forbid")

    normalized_name: str
    aliases: list[str]
    priority: Literal["must_do", "useful", "optional"]
    reason: str
    source_count: int = Field(ge=1)
    evidence: list[EvidenceReference] = Field(min_length=1)


class ActionPack(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str
    executive_summary: str
    backlog_reduction: BacklogReduction
    start_here: StartHere
    common_topics: list[CommonTopic]
    unique_insights: list[UniqueInsight]
    contradictions: list[Contradiction]
    resource_verdicts: list[ResourceVerdict]
    merged_notes: list[MergedNote]
    priority_problems: list[PriorityProblem]


class LearnerProfile(BaseModel):
    model_config = ConfigDict(extra="forbid")

    dsa_level: Literal["beginner", "intermediate", "advanced"] | None = None
    known_topics: list[str] = Field(default_factory=list, max_length=30)
    weak_topics: list[str] = Field(default_factory=list, max_length=30)
    preferred_language: str | None = Field(default=None, max_length=50)
    target_role_or_company: str | None = Field(default=None, max_length=120)
    available_study_minutes_per_day: int | None = Field(
        default=None,
        ge=10,
        le=1440,
    )
    target_interview_date: date | None = None


class ActionPackCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    learner_profile: LearnerProfile | None = None
    output_options: list[OutputOption] = Field(
        default_factory=lambda: ["complete_action_pack"],
        min_length=1,
        max_length=7,
    )
    visual_topics: list[str] = Field(default_factory=list, max_length=3)
    voice_mode: Literal["normal", "quick_revision"] = "normal"


class ActionPackResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: UUID
    project_id: UUID
    status: Literal["completed"]
    model: str
    source_ids: list[UUID]
    result_object_key: str
    generated_at: datetime
    action_pack: ActionPack
    output_options: list[OutputOption] = Field(
        default_factory=lambda: ["complete_action_pack"]
    )
    assets: list[GeneratedAsset] = Field(default_factory=list)

File: frontend/types/action-pack.ts
export interface EvidenceReference {
  resource_id: string;
  title: string;
  location: string | null;
  confidence: number;
  basis: "source_derived" | "ai_inferred";
  support: string;
}

export interface ResourceChoice {
  resource_id: string;
  title: string;
  reason: string;
  evidence: EvidenceReference[];
}

export interface ActionPack {
  title: string;
  executive_summary: string;
  backlog_reduction: {
    resource_count: number;
    estimated_original_minutes: number | null;
    repeated_content_percentage: number;
    metric_methodology: string;
    essential_resources: ResourceChoice[];
    optional_resources: ResourceChoice[];
    skippable_resources: ResourceChoice[];
  };
  start_here: {
    topic_or_resource: string;
    why: string;
    estimated_minutes: number | null;
    evidence: EvidenceReference[];
  };
  common_topics: Array<{
    topic: string;
    explanation: string;
    source_count: number;
    evidence: EvidenceReference[];
  }>;
  unique_insights: Array<{
    insight: string;
    why_it_matters: string;
    evidence: EvidenceReference[];
  }>;
  contradictions: Array<{
    topic: string;
    sides: Array<{ position: string; evidence: EvidenceReference[] }>;
    neutral_explanation: string;
    recommendation: string | null;
    recommendation_confidence: number | null;
  }>;
  resource_verdicts: Array<{
    resource_id: string;
    title: string;
    verdict:
      | "essential"
      | "use_selected_sections"
      | "reference_only"
      | "safe_to_skip"
      | "unavailable_or_low_confidence";
    reason: string;
    selected_sections: string[];
    evidence: EvidenceReference[];
  }>;
  merged_notes: Array<{
    topic: string;
    concise_notes: string[];
    syntax_or_pseudocode: string | null;
    recognition_clues: string[];
    common_mistakes: string[];
    memory_cues: string[];
    evidence: EvidenceReference[];
  }>;
  priority_problems: Array<{
    normalized_name: string;
    aliases: string[];
    priority: "must_do" | "useful" | "optional";
    reason: string;
    source_count: number;
    evidence: EvidenceReference[];
  }>;
}

export type OutputOption =
  | "complete_action_pack"
  | "quick_revision_notes"
  | "learning_workflow"
  | "voice_lesson"
  | "flashcards"
  | "priority_coding_problems"
  | "interview_revision_sheet";

export interface GeneratedAssetVersion {
  id: string;
  version_number: number;
  status: "generating" | "stored" | "failed";
  provider: string;
  model: string;
  mime_type: string;
  object_key: string | null;
  manifest_object_key: string | null;
  sha256: string | null;
  size_bytes: number | null;
  confidence: number | null;
  evaluation_summary: string | null;
  generation_time_ms: number | null;
  source_ids: string[];
  generation_settings: Record<string, unknown>;
  provenance: {
    classification: "source_derived" | "ai_generated" | "ai_inferred";
    resources: Array<{
      resource_id: string;
      title: string;
      link: string | null;
    }>;
    evidence_references: Array<Record<string, unknown>>;
    generation_timestamp: string;
    version_number: number;
  };
  genblaze_run_id: string | null;
  parent_version_number: number | null;
  failure_message: string | null;
  created_at: string;
  is_current: boolean;
  download_url: string | null;
}

export interface GeneratedAsset {
  id: string;
  action_pack_id: string;
  project_id: string;
  asset_type:
    | "complete_action_pack"
    | "note"
    | "learning_workflow"
    | "voice"
    | "flashcards"
    | "priority_problems"
    | "interview_revision_sheet";
  logical_key: string;
  display_name: string;
  current_version_number: number | null;
  created_at: string;
  updated_at: string;
  versions: GeneratedAssetVersion[];
}

export interface LearningWorkflowStage {
  stage_id: string;
  label: string;
  headline: string;
  summary: string;
  tone: "teal" | "blue" | "violet" | "amber" | "coral" | "green" | "navy";
  items: string[];
  evidence: Array<Record<string, unknown>>;
  estimated_minutes: number | null;
}

export interface LearningWorkflowAsset {
  schema_version: 1;
  action_pack_id: string;
  project_id: string;
  title: string;
  summary: string;
  mode: "guided" | "concise";
  focus_topics: string[];
  source_ids: string[];
  stages: LearningWorkflowStage[];
}

export interface ActionPackResponse {
  id: string;
  project_id: string;
  status: "completed";
  model: string;
  source_ids: string[];
  result_object_key: string;
  generated_at: string;
  action_pack: ActionPack;
  output_options: OutputOption[];
  assets: GeneratedAsset[];
}
